### ISSUE TITLE
Index corruption when upgrading from macOS 10.15 or earlier to macOS 11 or later

### ISSUE THREAD

**Issue by jakob** [PostgresApp/PostgresApp#665]
@tbussmann has discovered an issue where indexes are corrupted when upgrading from a system running macOS 10.15 or earlier to macOS 11 or later.

The underlying issue is that Apple has changed some system collations in macOS 11, including the en_US.utf-8 collation that Postgres.app uses by default.

This problem can lead to the following problems:

- SELECT or UPDATE queries could miss some rows
- UNIQUE constraints might not be correctly enforced, leading to duplicate keys

To fix the issue, you have to reindex every database of the server. To do so, connect to each database and execute the following SQL query:

```sql
REINDEX DATABASE <database_name>;
```

The query can take anywhere from a few seconds to minutes or hours, depending on the size of your database.

The query will fail if you already have duplicate keys in a table. If that is the case, PostgreSQL will show a warning telling you about the problem. Identify the problematic table, fix the duplicate keys (be careful because queries might fail to show all rows because of the corrupted index), then try again.

**Comment by jakob** [comment]
We are planning to have Postgres.app automatically show a warning if this issue could affect you. Our plan is to do the following:

- when data directories are initialized, store the current macOS version and a hash or some other identifier of the default collation
- when running Postgres.app, we check these parameters for compatibility, and if we detect a problem we show a warning

- when the user dismisses the warning, we save the macOS version and hash again, and compare to that instead

But this only works for new data directories. For existing data directories, we try to guess which macOS version it was created on by comparing the creation date of the pg_version file to installation dates of macOS updates extracted from /Library/Receipts

**Comment by tbussmann** [comment]
Some additional details on when this may be (not) need:
You are not affected if:
* the database cluster ('Server' in PostgresApp) was initialised on macOS 11 or later
* you are still using macOS 10.15 or earlier 
* `initdb` was run manually (not with the button 'Initialize' within PostgresApp) with `--lc-collate` or `--locale` set to "C" or "POSIX" or `--no-locale` and no other libc-based collations are used on columns or in indexes. Database default collation can be shown with `SELECT datname, datcollate FROM pg_database;`
* the database was restored from a logic dump (`pg_dump` / `pg_dumpall`) after updating the OS version

You are likely affected if:
* the the OS was updated from macOS 10.15 or earlier to macOS 11 or later 'in-place' with an existing PostgreSQL database
* you used Migration Assistant or TimeMachine to copy the user data containing a PostgreSQL from a machine using macOS 10.15 or earlier to macOS 11 or later
* you used `pg_upgrade` on  macOS 11 or later to update a cluster inited on macOS 10.15 or earlier
* you copied over a data directory or a physical (`pg_basebackup`) from a machine using macOS 10.15 or earlier to macOS 11 or later

There is no correlation to the version of PostgresApp or PostgreSQL or on the architecture (Apple Silcon / Intel)

**Comment by jakob** [comment]
As of commit 5c96c8ccde4bfa5fcfe8da807a135019efd17c08, Postgres.app now tries to warn the user to execute REINDEX when necessary.

To do so, it records the version of macOS where initdb was called, and also all versions of macOS that the server was started on. This information is stored in the data directory in a file named "postgresapp_config.plist".

Since this information is not available for old data directories, Postgres.app guesses the macOS version used to call initdb for existing data directories. This check is performed the first time the new version of Postgres.app is launched.

If Postgres.app detects that the data directory was used both pre macOS 11 and post macOS 11, it shows the reindex warning.

If the macOS version used for initdb of a data directory is unknown (eg. restored from backup, moved with migration assistant), then Postgres.app also shows the warning.

I'll try to upload a build for testing next week.

**Comment by jakob** [comment]
I've updated the "reindex-warning" page to be more generic. It now covers PostgreSQL bugs as well as macOS collation changes.

I've also changed the instructions to tell the user to connect to the database and execute the "REINDEX DATABASE dbname;" command. I think this is better for two reasons:

1. It's the same method suggested in the reindex warning
2. Some users may not be familiar with the command line because they use graphical clients. However, almost everyone should know how to execute SQL commands on their database.
